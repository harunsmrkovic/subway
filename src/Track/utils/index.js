// Determines positions for railways between two stations
// Will shift each railway for given interspace from the center line between stations

export const determineRailsPosition = (
  fromX,
  fromY,
  toX,
  toY,
  railsInterspace
) => {
  const halfDistance = railsInterspace / 2
  const angleCorrection = railsInterspace * 0.15

  // Horizontal rails
  if (fromY === toY) {
    const lineA = {
      fromX,
      fromY: fromY - halfDistance,
      toX,
      toY: toY - halfDistance
    }
    const lineB = {
      fromX,
      fromY: fromY + halfDistance,
      toX,
      toY: toY + halfDistance
    }

    return fromX < toX ? [lineA, lineB] : [lineB, lineA]
  }

  // Vertical rails
  if (fromX === toX) {
    const lineA = {
      fromX: fromX - halfDistance,
      fromY,
      toX: toX - halfDistance,
      toY
    }
    const lineB = {
      fromX: fromX + halfDistance,
      fromY,
      toX: toX + halfDistance,
      toY
    }

    return fromY > toY ? [lineA, lineB] : [lineB, lineA]
  }

  // Diagonal rails
  if (fromY < toY && fromX < toX) {
    // Top-Left to Bottom-Right
    return [
      {
        fromX: fromX + halfDistance - angleCorrection,
        fromY: fromY - halfDistance + angleCorrection,
        toX: toX + halfDistance - angleCorrection,
        toY: toY - halfDistance + angleCorrection
      },
      {
        fromX: fromX - halfDistance + angleCorrection,
        fromY: fromY + halfDistance - angleCorrection,
        toX: toX - halfDistance + angleCorrection,
        toY: toY + halfDistance - angleCorrection
      }
    ]
  } else if (fromX > toX && fromY > toY) {
    // Bottom-Right to Top-Left
    return [
      {
        fromX: fromX - halfDistance + angleCorrection,
        fromY: fromY + halfDistance - angleCorrection,
        toX: toX - halfDistance + angleCorrection,
        toY: toY + halfDistance - angleCorrection
      },
      {
        fromX: fromX + halfDistance - angleCorrection,
        fromY: fromY - halfDistance + angleCorrection,
        toX: toX + halfDistance - angleCorrection,
        toY: toY - halfDistance + angleCorrection
      }
    ]
  } else {
    // Bottom-Left to Top-Right
    const lineA = {
      fromX: fromX - halfDistance + angleCorrection,
      fromY: fromY - halfDistance + angleCorrection,
      toX: toX - halfDistance + angleCorrection,
      toY: toY - halfDistance + angleCorrection
    }
    const lineB = {
      fromX: fromX + halfDistance - angleCorrection,
      fromY: fromY + halfDistance - angleCorrection,
      toX: toX + halfDistance - angleCorrection,
      toY: toY + halfDistance - angleCorrection
    }

    return fromX < toX && fromY > toY ? [lineA, lineB] : [lineB, lineA]
  }
}
