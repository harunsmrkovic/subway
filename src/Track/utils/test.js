import { determineRailsPosition } from '.'

const railsInterspace = 3

describe('Track utils', () => {
  describe('determineRailsPosition', () => {
    describe('Diagonal', () => {
      describe('Bottom-Left towards Top-Right', () => {
        it('distances both lines correctly', () => {
          expect(
            determineRailsPosition(5, 10, 10, 5, railsInterspace)
          ).toMatchSnapshot()
        })
      })

      describe('Top-Left towards Bottom-Right', () => {
        it('distances both lines correctly', () => {
          expect(
            determineRailsPosition(90, 5, 95, 10, railsInterspace)
          ).toMatchSnapshot()
        })
      })

      describe('Top-Right towards Bottom-Left', () => {
        it('distances both lines correctly', () => {
          expect(
            determineRailsPosition(95, 50, 90, 55, railsInterspace)
          ).toMatchSnapshot()
        })
      })

      describe('Bottom-Right towards Top-Left', () => {
        it('distances both lines correctly', () => {
          expect(
            determineRailsPosition(10, 55, 5, 50, railsInterspace)
          ).toMatchSnapshot()
        })
      })
    })

    describe('Vertical', () => {
      describe('Top towards Bottom', () => {
        it('distances both lines correctly', () => {
          expect(
            determineRailsPosition(95, 10, 95, 50, railsInterspace)
          ).toMatchSnapshot()
        })
      })

      describe('Bottom towards Top', () => {
        it('distances both lines correctly', () => {
          expect(
            determineRailsPosition(5, 50, 5, 15, railsInterspace)
          ).toMatchSnapshot()
        })
      })
    })

    describe('Horizontal', () => {
      describe('Left towards Right', () => {
        it('distances both lines correctly', () => {
          expect(
            determineRailsPosition(10, 5, 90, 5, railsInterspace)
          ).toMatchSnapshot()
        })
      })

      describe('Right towards Left', () => {
        it('distances both lines correctly', () => {
          expect(
            determineRailsPosition(90, 55, 10, 55, railsInterspace)
          ).toMatchSnapshot()
        })
      })
    })
  })
})
