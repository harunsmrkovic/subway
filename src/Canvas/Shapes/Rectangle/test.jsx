import React from 'react'
import { shallow } from 'enzyme'
import { fingerprint } from '../../../tests/utils'
import { RectangleBase as Rectangle } from '.'

const defaultProps = {
  width: 10,
  height: 10,
  x: 1,
  y: 1
}

describe('Rectangle', () => {
  describe('radius', () => {
    describe('when it has no corner radius', () => {
      it('uses provided ctx to draw Rectangle with no corner radius', () => {
        const ctx = fingerprint([
          'beginPath',
          'moveTo',
          'lineTo',
          'stroke',
          'closePath'
        ])

        shallow(<Rectangle {...defaultProps} ctx={ctx.mocks} />)
        expect(ctx.log).toMatchSnapshot()
      })
    })
  })
})
