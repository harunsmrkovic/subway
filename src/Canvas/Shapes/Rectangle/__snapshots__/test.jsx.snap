// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rectangle radius when it has no corner radius uses provided ctx to draw Rectangle with no corner radius 1`] = `
Array [
  Object {
    "args": Array [],
    "fn": "beginPath",
  },
  Object {
    "args": Array [
      1,
      1,
    ],
    "fn": "moveTo",
  },
  Object {
    "args": Array [
      11,
      1,
    ],
    "fn": "lineTo",
  },
  Object {
    "args": Array [
      11,
      11,
    ],
    "fn": "lineTo",
  },
  Object {
    "args": Array [
      1,
      11,
    ],
    "fn": "lineTo",
  },
  Object {
    "args": Array [
      1,
      1,
    ],
    "fn": "lineTo",
  },
  Object {
    "args": Array [],
    "fn": "stroke",
  },
  Object {
    "args": Array [],
    "fn": "closePath",
  },
]
`;
